op-service, op-batcher, op-proposer: Active sequencer follow mode

EvanJRichard opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

This PR proposes to add `dial.ActiveL2EndpointProvider` and `dial.ActiveL2RollupProvider`. These endpoint providers will check that they are pointing to the active sequencer before providing rollup clients and l2 ethClients to callers. They are enabled by passing a comma-separated list as the rollup/l2 ethclient URL string. Additionally, this PR adds infrastructure for unit testing these structs, such as `testutils.MockRollupClient`.

**Tests**
See new unit test file, `op-service/dial/active_l2_provider_test.go`, about half the changeset is that file. Most unit tests follow the pattern where an `ActiveL2RollupProvider`/`ActiveL2EndpointProvider` is pointing at an active sequencer, "notices" that the sequencer goes inactive, and then dials a new sequencer that reports it is active. 

This also successfully failed over in a devnet-like test run locally on my machine. It has also been tested to work on our internal cloud devnet during a ZDD.

**Metadata**
- Requires and builds on top of #8580 
- Fixes https://github.com/ethereum-optimism/client-pod/issues/166

Commit:
op-service: Add ActiveL2EndpointProvider.

Commit:
Fix bug in initialization, and handle case where no ethUrls are provided.

Commit:
Split active L2 provider into active rollup and active L2 provider.

Commit:
Re-duplicate some code until tests are passing.

Commit:
op-proposer: Add ability to enable active provider.

Commit:
op-batcher: Add ability to enable active provider.

Commit:
Add an empty test skeleton.

Commit:
Add an empty test skeleton.

Commit:
Merge branch 'evan/batcher-active-seq' of github.com:ethereum-optimism/optimism into evan/batcher-active-seq

Commit:
op-service: add, but do not yet use, RollupClientInterface and EthClientInterface.

Commit:
op-service: update mocks and interfaces for endpoint provider testing.

Commit:
op-service - WIP on Active L2 Providers: unit tests pass, design and impl contains TODOs.

Commit:
op-service: restore design in Active Endpoint Providers that only keeps one client open at a time.

Commit:
op-service: when dialing a new sequencer, close() the old connection.

EvanJRichard commented on op-batcher/batcher/config.go:
```diff
@@ -20,10 +20,10 @@ type CLIConfig struct {
 
-	// L2EthRpc is the HTTP provider URL for the L2 execution engine.
+	// L2EthRpc is the HTTP provider URL for the L2 execution engine. A comma-separated list enables the active L2 provider. Such a list needs to match the number of RollupRpcs provided.
 	L2EthRpc string
 
-	// RollupRpc is the HTTP provider URL for the L2 rollup node.
+	// RollupRpc is the HTTP provider URL for the L2 rollup node. A comma-separated list enables the active L2 provider. Such a list needs to match the number of L2EthRpcs provided.
 	RollupRpc string
```
This is checked on provider creation.

EvanJRichard commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,131 @@
+	return nil
+}
+
+func (p *ActiveL2EndpointProvider) findActiveEndpoints(ctx context.Context) error {
+	const maxRetries = 20
+	totalAttempts := 0
+
+	for totalAttempts < maxRetries {
+		active, err := p.checkCurrentSequencer(ctx)
+		if err != nil {
+			p.log.Warn("Error querying active sequencer, closing connection and trying next.", "err", err, "try", totalAttempts)
+			p.currentRollupClient.Close()
+			p.currentEthClient.Close()
+		} else if active {
+			p.log.Debug("Current sequencer active.", "try", totalAttempts)
+			return nil
+		} else {
+			p.log.Info("Current sequencer inactive, closing connection and trying next.", "try", totalAttempts)
+			p.currentRollupClient.Close()
+			p.currentEthClient.Close()
+		}
+		if err := p.dialNextSequencer(ctx); err != nil {
+			return fmt.Errorf("dialing next sequencer: %w", err)
+		}
+
+		totalAttempts++
+		if p.currentIndex >= p.numEndpoints() {
+			p.currentIndex = 0
+		}
+	}
+	return fmt.Errorf("failed to find an active sequencer after %d retries", maxRetries)
```
We will leave it up to callers to call `Close()` in this case.

Commit:
op-service: obey coderabbit suggestion around safer handling of p.currentIndex in Active L2 Providers.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,128 @@
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+	ethDialer ethDialer,
+	rollupDialer rollupDialer,
```
I suggest we make `NewActiveL2EndpointProvider` private and then do something like
```go
func NewActiveL2EndpointProvider(
	ctx context.Context,
	ethUrls, rollupUrls []string,
	checkDuration time.Duration,
	networkTimeout time.Duration,
	logger log.Logger) {
    return newActiveL2EndpointProvider(ctx, ethUrls, checkDuration, networkTimeout, logger, 
dial.DialEthClientInterfaceWithTimeout, dial.DialRollupClientInterfaceWithTimeout)
}
```
I would also just inline the two functions `dial.DialEthClientInterfaceWithTimeout, dial.DialRollupClientInterfaceWithTimeout`, because, IIUC, they are only needed here.

This way, we can use the `newActiveL2EndpointProvider` in tests and have a nicer API for `NewActiveL2EndpointProvider`, and can also remove the two public functions `dial.DialEthClientInterfaceWithTimeout` and `dial.DialRollupClientInterfaceWithTimeout`.

Similarly for `NewActiveL2RollupProvider`.

sebastianst replied:
Note also that public functions shouldn't receive non-exported types.

EvanJRichard replied:
That cleans up the external calls massively, great idea.

sebastianst commented on op-service/testutils/mock_rollup_client.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func (m *MockRollupClient) SequencerActive(ctx context.Context) (bool, error) {
+	out := m.Mock.MethodCalled("SequencerActive")
+	return out[0].(bool), *out[1].(*error)
```
```suggestion
	out := m.Mock.Called()
	return out.Bool(0), out.Error(1)
```
* `mock` uses reflection to get the method name, should work here
* In general, all arguments are passed to `Called`, like here `Called(ctx)`. I can see how in this instance, we never care about the `ctx` value, so might as well skip like you do already 👍🏻 
* `mock` provides convenience return methods for bools, ints, errors etc
* I think it should just be possible to use normal `error`s, without pointers. Then you also have to change the `ExpectSequencerActive` definition to just return `active, err`.

Similarly in all other mock definitions

EvanJRichard replied:
Cool. I might take some time after this PR is done, then, to fix up the other `mock_`s too, I was over-matching their style

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,138 @@
+	const maxRetries = 20
+	totalAttempts := 0
+
+	for totalAttempts < maxRetries {
```
why not a regular
```suggestion
	for totalAttempts := 0; totalAttempts < maxRetries; totalAttempts++ {
```
?
(Then have to remove the extra `totalAttempts++`)

EvanJRichard replied:
Agreed. Weird pattern that snuck in during a debugging session. Will fix.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,138 @@
+	return time.Now().After(p.activeTimeout)
+}
+
+func (p *ActiveL2RollupProvider) findActiveEndpoints(ctx context.Context) error {
```
I see two issues here
* `maxRetries` is the total number of individual attempts, not number of attempts over the whole list of endpoints. E.g. if you have 3 endpoints, you'd try the first two 7 times and the last one 6 times. I think it would be cleaner to have `maxRetries` mean the number of attempts to go over the whole list.
* You don't wait between attempts over the whole list. So you would try all offline sequencers in very fast succession, probably not leading to success. If a sequencer is offline, it probably will be 10ms later. It would be better to wait after having done one full iteration over the endpoint list.

You could fix both by using [`retry.Do`](https://github.com/ethereum-optimism/optimism/blob/6e371b4bc31b99f726f439da8b82186d2e084475/op-service/retry/operation.go#L41) with the `Fixed` strategy. As `op` function, you then pass a single loop over all endpoints.

A simple and less complex alternative could be to just iterate once over the list, and don't have multiple attempts over the list at all, removing the concept of `maxRetries`. This could work in that the callers are then driving the retrying. Probably enough as a first simple solution.

EvanJRichard replied:
I really like that `retry.Do`, but you're right, to start I'll just to a one-iteration-over-the-list design and let callers drive retrying.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,128 @@
+	return nil
+}
+
+func (p *ActiveL2EndpointProvider) findActiveEndpoints(ctx context.Context) error {
```
This is almost completely duplicating the logic of `ActiveL2RollupProvider.findActiveEndpoints` by the embedded `ActiveL2RollupProvider`. So if we make a change there, we have to carefully apply it here.

I'm wondering if there's a way to reuse `ActiveL2RollupProvider.findActiveEndpoints` by just calling it here, and then on success dialing the `EthClient` for the active index that has been determined. You could then also throw away the duplicated `dialNextSequencer`.

EvanJRichard replied:
Yeah, I was on the fence about duplicating the code vs. "doing the same thing in two places isn't enough to worry about DRY yet". But it's weird that some of the code is duplicated but some uses the embedded logic. I'll find a way to try and make it almost-entirely-embedded as you suggest in the second paragraph there, there's gotta be a way. EDIT: Yeah, the way you describe works great, implementing.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,104 @@
+	// Check that the first client is used, then the second once the first declares itself inactive.
+	firstSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.True(t, &primarySequencer == firstSequencerUsed) // avoids copying the struct (and its mutex, etc.)
```
I suggest to use [`require.Same`](https://pkg.go.dev/github.com/stretchr/testify/require#Same) here.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,104 @@
+	require.True(t, &primarySequencer == firstSequencerUsed) // avoids copying the struct (and its mutex, etc.)
+	secondSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.True(t, &secondarySequencer == secondSequencerUsed)
```
I suggest to use [`require.Same`](https://pkg.go.dev/github.com/stretchr/testify/require#Same) here and in all other pointer-equality assertions.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,104 @@
+// will failover to the next provider if the current one is not active.
+func TestActiveSequencerFailoverBehavior_RollupProviders(t *testing.T) {
+	// Create two mock rollup clients, one of which will declare itself inactive after first check.
+	primarySequencer := testutils.MockRollupClient{}
```
maybe just do
```suggestion
	primarySequencer := new(testutils.MockRollupClient)
```
so you can naturally treat them as pointers down below and don't have to always get the reference.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,104 @@
+
+// TestActiveSequencerFailoverBehavior_RollupProvider tests that the ActiveL2RollupProvider
+// will failover to the next provider if the current one is not active.
+func TestActiveSequencerFailoverBehavior_RollupProviders(t *testing.T) {
```
A second test would be nice where both sequencers are unreachable like 2 times and then only on the third attempt one is online.

However, maybe we also decide to simplify the impl to not run multiple attempts over the list at all.

EvanJRichard replied:
Resolved: new design (for now) only checks the whole list once.

sebastianst commented on op-batcher/batcher/service.go:
```diff
@@ -125,9 +126,16 @@ func (bs *BatcherService) initRPCClients(ctx context.Context, cfg *CLIConfig) er
 
-	endpointProvider, err := dial.NewStaticL2EndpointProvider(ctx, bs.Log, cfg.L2EthRpc, cfg.RollupRpc)
+	var endpointProvider dial.L2EndpointProvider
+	if strings.Contains(cfg.RollupRpc, ",") || strings.Contains(cfg.L2EthRpc, ",") {
```
Shouldn't this be 
```suggestion
	if strings.Contains(cfg.RollupRpc, ",") && strings.Contains(cfg.L2EthRpc, ",") {
```
I mean, if only one is a list, it's a wrong config anyways.

Should also add a check to `CLIConfig.Check` that both are lists of the same length.

EvanJRichard replied:
My thinking was that if just one list is provided, we should avoid silently falling back to the `Static` provider. Adding the check to `CLIConfig.Check` is a better way of achieving this. :+1: (Edit: there, though, the check _will_ be `||` so that it will length-check and then error in the just-one-list provided case.)

sebastianst reviewed (dismissed):
Going in a good direction. Maybe we want to simplify further by not doing retries after having iterated over the whole list once. We then offload the retry driving to the callers. Might work well, should check.

Commit:
op-service, op-batcher, op-proposer: address review comments in PR#8585.

Commit:
op-service: Active L2 Provider - add test case for a sequencer returning an error.

Commit:
op-service: Active L2/Rollup Providers: improve unit testing and logging.

sebastianst commented on op-batcher/batcher/config.go:
```diff
@@ -74,6 +75,13 @@ func (c *CLIConfig) Check() error {
+		if len(rollupUrls) != len(ethUrls) {
+			return errors.New("number of rollup and eth URLs must match")
+		}
+	}
```
You can shorten this to
```suggestion
	if strings.Count(c.RollupRpc, ",") != strings.Count(c.L2EthRpc, ",") {
		return errors.New("number of rollup and eth URLs must match")
	}
```

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,94 @@
+func (p *ActiveL2EndpointProvider) EthClient(ctx context.Context) (EthClientInterface, error) {
+	p.clientLock.Lock()
+	defer p.clientLock.Unlock()
+	indexBeforeCheck := p.currentIndex
```
I think this doesn't work and we need to track the `EthClient` index separately. If you called `RollupClient` before and that changed the index, we wouldn't notice here, because the `currentIndex` tracks the rollup client, and then the `ensureActiveEndpoint` wouldn't change it again and we wouldn't notice that the `currentEthClient` is from a different index.

 E.g. rename `currentIndex` to `rollupIndex` and add a 2nd `ethClientIndex` field to the `ActiveL2EndpointProvider` and then first call `ensureActiveEndpoint` and then compare rollup and ethclient indices.

Ideally add a failing test first that verifies my suspicion that if the rollup client changes the index, we wouldn't notice here, then fix it.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,94 @@
+	if err != nil {
+		return nil, err
+	}
+	if indexBeforeCheck != p.currentIndex {
```
Here you'd then do
```suggestion
	if p.ethClientIndex != p.rollupIndex {
```

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,147 @@
+			p.log.Debug("Current sequencer active.", "index", index, "url", ep)
+			return nil
+		} else {
+			p.log.Info("Current sequencer inactive, closing connection and trying next.", "index", index, "url", ep)
```
The "index" in these logs is misleading, because it's not the absolute index, but the _offset_ to the initial `currentIndex`.

I suggest just using `p.currentIndex` as the value for the "index" log field, and rename the `index` variable to `offset`. And now it looks like we don't need the offset at all, so you can just do `for range ...` and drop the loop var.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,94 @@
+	ethClient, err := ethDialer(cctx, networkTimeout, logger, ethUrls[0])
+	if err != nil {
+		return nil, fmt.Errorf("dialing eth client: %w", err)
+	}
```
This now has the implicit assumption that the first ethclient is available when creating the `ActiveL2EndpointProvider`. I think we need to be more robust here and already find the active one. It may well happen in production situations that a batcher is restarted while the first sequencers is inactive, or even some offline, e.g. during incidents, so being robust here is important.

Should be possible without code duplication by employing the already existing logic like
```go
ep :=  &ActiveL2EndpointProvider{ /* already set all fields but the `currentEthClient` */ }
_, err = ep.EthClient(ctx)
if err != nil {
    //...
}
return ep, nil
```
For this, you probably need to add logic that deals with a `nil` `currentEthClient`.

We should add a test that checks that the constructor can deal with a situation where the first endpoint is offline or inactive.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,147 @@
+	rollupClient, err := dialer(cctx, networkTimeout, logger, rollupUrls[0])
+	if err != nil {
+		return nil, fmt.Errorf("dialing rollup client: %w", err)
+	}
```
This now has the implicit assumption that the first rollupclient is available when creating the `ActiveL2RollupProvider`. I think we need to be more robust here and already find the active one. It may well happen in production situations that a batcher is restarted while the first sequencers is inactive, or even some offline, e.g. during incidents, so being robust here is important.

Should be possible without code duplication by employing the already existing logic like
```go
rp :=  & ActiveL2RollupProvider{ /* already set all fields but the `currentRollupClient` */ }
_, err = rp.RollupClient(ctx)
if err != nil {
    //...
}
return rp, nil
```
For this, you probably need to add logic that deals with a `nil` `currentRollupClient`.

We should add a test that checks that the constructor can deal with a situation where the first endpoint is offline or inactive.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,147 @@
+	networkTimeout time.Duration,
+	logger log.Logger,
+) (*ActiveL2RollupProvider, error) {
+	dialRollupClientInterfaceWithTimeout := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
```
nit: it's a local var and very long line, maybe rename to smth shorter like `rollupDialer` and add line breaks to the function sig.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,195 @@
+	// as TestActiveSequencerFailoverBehavior_RollupProviders,
+	// but ensure the added `EthClient()` method also triggers the failover.
+	primarySequencer := new(testutils.MockRollupClient)
+	primarySequencer.ExpectSequencerActive(true, nil)
+	primarySequencer.ExpectSequencerActive(false, nil)
+	primarySequencer.ExpectClose()
+	secondarySequencer := new(testutils.MockRollupClient)
+	secondarySequencer.ExpectSequencerActive(true, nil)
+
+	mockRollupDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
+		if url == "primaryRollup" {
+			return primarySequencer, nil
+		} else if url == "secondaryRollup" {
+			return secondarySequencer, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+	primaryEthClient := new(testutils.MockEthClient)
+	primaryEthClient.ExpectClose()
+	secondaryEthClient := new(testutils.MockEthClient)
+	mockEthDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (EthClientInterface, error) {
+		if url == "primaryEth" {
+			return primaryEthClient, nil
+		} else if url == "secondaryEth" {
+			return secondaryEthClient, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+	endpointProvider, err := newActiveL2EndpointProvider(
+		context.Background(),
+		[]string{"primaryEth", "secondaryEth"},
+		[]string{"primaryRollup", "secondaryRollup"},
+		1*time.Microsecond,
+		1*time.Minute,
+		testlog.Logger(t, log.LvlDebug),
+		mockEthDialer,
+		mockRollupDialer,
+	)
```
A lot of the setup code could is duplicated in tests. I think adding a test harness could be beneficial here and make it easier to write more tests, while make the existing ones more readable, like
```go
type endpointProviderTest struct {
    rollupClients []*testutils.MockRollupClient
    ethClients []*testutils.MockEthClient
    // dialers, ...
}

func setupEndpointProviderTest(t * testing.T, numSequencers int) *endpointProviderTest {
    // all the setup, except setting up the mocks
}
```

to make it possible to setup the mocks before calling the constructor, you could add a method to the harness like
```go
func (et *endpointProviderTest) newActiveL2EndpointProvider() (*ActiveL2EndpointProvider, error)
```

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,195 @@
+
+// TestActiveSequencerFailoverBehavior_RollupProviders_Inactive tests that the ActiveL2RollupProvider
+// will failover to the next provider if the current one is not active.
+func TestActiveSequencerFailoverBehavior_RollupProviders_Inactive(t *testing.T) {
+	// Create two mock rollup clients, one of which will declare itself inactive after first check.
+	primarySequencer := new(testutils.MockRollupClient)
+	primarySequencer.ExpectSequencerActive(true, nil)
+	primarySequencer.ExpectSequencerActive(false, nil)
+	primarySequencer.ExpectClose()
+	secondarySequencer := new(testutils.MockRollupClient)
+	secondarySequencer.ExpectSequencerActive(true, nil)
+
+	mockRollupDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
+		if url == "primaryRollup" {
+			return primarySequencer, nil
+		} else if url == "secondaryRollup" {
+			return secondarySequencer, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+
+	endpointProvider, err := newActiveL2RollupProvider(
+		context.Background(),
+		[]string{"primaryRollup", "secondaryRollup"},
+		1*time.Microsecond,
+		1*time.Minute,
+		testlog.Logger(t, log.LvlDebug),
+		mockRollupDialer,
+	)
+	require.NoError(t, err)
+	// Check that the first client is used, then the second once the first declares itself inactive.
+	firstSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, firstSequencerUsed)
+	secondSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, secondarySequencer, secondSequencerUsed)
```
I think the rabbit is right here, should add a test that asserts that getting a rollup or ethclient fails if all are offline or inactive.
We should also add a few more tests like
* that it just returns the first one if it is active and that it then doesn't iterate.
* that if it fails over the the second one, and then the second one becomes inactive but the first one active, that it then goes back to the first one.
* that constructors find the active sequencer at the time of creation (see other comment around this)
* that constructor fails if all are offline
* etc

sebastianst replied:
Also need a test that if the check duration hasn't passed yet, that it then just returns the latest ethclient and rollup clients without checking.

sebastianst reviewed (changes requested):
The ActiveRollupProvider is currently implicitly tested via the ActiveEndpointProvider tests. This is fine, but we should double check that we don't miss a code path from it in tests.

Commit:
op-service, op-batcher: address review comments in 8585 regarding first-startup behavior and testing.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,153 @@
+		active, err := p.checkCurrentSequencer(ctx)
+		ep := p.rollupUrls[p.rollupIndex]
+		if err != nil {
+			p.log.Warn("Error querying active sequencer, closing connection and trying next.", "err", err, "offset", p.rollupIndex, "url", ep)
```
The `rollupIndex` is an absolute index. Only the loop var was an offset.
```suggestion
			p.log.Warn("Error querying active sequencer, closing connection and trying next.", "err", err, "index", p.rollupIndex, "url", ep)
```

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,153 @@
+		if err != nil {
+			p.log.Warn("Error querying active sequencer, closing connection and trying next.", "err", err, "offset", p.rollupIndex, "url", ep)
+		} else if active {
+			p.log.Debug("Current sequencer active.", "offset", p.rollupIndex, "url", ep)
```
"index"

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,153 @@
+			p.log.Debug("Current sequencer active.", "offset", p.rollupIndex, "url", ep)
+			return nil
+		} else {
+			p.log.Info("Current sequencer inactive, closing connection and trying next.", "offset", p.rollupIndex, "url", ep)
```
"index"

ajsutton reviewed (commented):
Generally looks good - agree with Seb about the tests and will leave final approval for him.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,153 @@
+			return nil, fmt.Errorf("dialing rollup client: %w", err)
+		}
+		p.currentRollupClient = rollupClient
+	}
```
ok I see, this is needed to initialize the field. Since the only code path that leads to a `nil` `currentRollupClient` is right after construction, I propose we actually move this back into `newActiveL2RollupProvider` right before calling `_, err := rp.RollupClient(cctx)` afterwards. Gives us the best of both worlds, clean `RollupClient()` and an active client to start with after construction.

My initial idea was that we somehow handle the nil case in `dialNextSequencer` to completely deduplicate dialing code, but this also works fine.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,104 @@
+			p.currentEthClient.Close()
+		}
+		p.currentEthClient = ethClient
+	}
```
ok I see, this is needed to initialize the field. Since the only code path that leads to a `nil` `currentEthClient` is right after construction, I propose we actually move this back into `newActiveL2EndpointProvider` right before calling `_, err := rp. EthClient(cctx)` afterwards. Gives us the best of both worlds, clean `EthClient()` and an active client to start with after construction. 

Here, we might actually return an error straight away in the ctor if dialing for the index `p.ethClientIndex = p.rollupIndex` fails because the ctor calls `newActiveL2RollupProvider` and if the EL client for the just determined index is offline, it's imo an error.

My initial idea was that we somehow handle the nil case in `dialNextSequencer` to completely deduplicate dialing code, but this also works fine.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,468 @@
+}
+
+// newActiveL2EndpointProvider constructs a new ActiveL2RollupProvider using the test harness setup.
+func (et *endpointProviderTest) newActiveL2RollupProvider(checkDuration time.Duration) (*ActiveL2RollupProvider, error) {
```
nice, love how this deduplicates stuff

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,468 @@
+	primarySequencer.ExpectSequencerActive(true, nil) // respond true again when the test calls `RollupClient()` the first time
+
+	shortCheckDuration := 1 * time.Microsecond
+	rollupProvider, err := ept.newActiveL2RollupProvider(shortCheckDuration)
```
if you want to just always have it check in tests, I think you can just use `0` duration
```suggestion
	rollupProvider, err := ept.newActiveL2RollupProvider(0)
```

you could also add to the doc of the ActiveRollupProvider that 0 enables check on every call to `RollupClient`.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,468 @@
+
+// TestRollupProvider_InitialActiveSequencerSelection verifies that the ActiveL2RollupProvider
+// selects the active sequencer correctly at the time of creation.
+func TestRollupProvider_InitialActiveSequencerSelection(t *testing.T) {
```
we should also add tests that if seq 0 is inactive at creation, but seq 1 is active, seq 1 is chosen from the beginning.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,468 @@
+}
+
+// TestRollupProvider_LongCheckDuration verifies the behavior of ActiveL2RollupProvider with a long check duration.
+func TestRollupProvider_LongCheckDuration(t *testing.T) {
```
could also add a test that even if sequencer is inactive, still the same is returned because the check is skipped.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,468 @@
+
+// TestEndpointProvider_FailOnAllErroredSequencers verifies that the ActiveL2EndpointProvider
+// fails to create when all sequencers return an error.
+func TestEndpointProvider_FailOnAllErroredSequencers(t *testing.T) {
```
I think we're still missing a test that if we can successfully create the provider, but then all are inactive or borked, `EthClient()` and `RollupClient()` return an error.

sebastianst reviewed (changes requested):
Looking pretty good now! Just a couple minor remarks and more tests.

Commit:
op-service: address review comments through adding more tests, and moving "nil client" behavior from client getter to constructor.

Commit:
op-service: minor error message change in active endpoint providers.

EvanJRichard committed:
Update op-service/dial/active_l2_provider.go

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

EvanJRichard commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,575 @@
+}
+
+// TestRollupProvider_LongCheckDuration verifies the behavior of ActiveL2RollupProvider with a long check duration.
+func TestRollupProvider_LongCheckDuration(t *testing.T) {
+	ept := setupEndpointProviderTest(t, 2)
+	primarySequencer := ept.rollupClients[0]
+
+	longCheckDuration := 1 * time.Hour
+	primarySequencer.ExpectSequencerActive(true, nil) // Active check on creation
+
+	rollupProvider, err := ept.newActiveL2RollupProvider(longCheckDuration)
+	require.NoError(t, err)
+
+	// Should return the same client without extra checks
+	firstSequencerUsed, err := rollupProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, firstSequencerUsed)
+
+	secondSequencerUsed, err := rollupProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, secondSequencerUsed)
+}
```
I believe `TestRollupProvider_ReturnsSameSequencerOnInactiveWithLongCheckDuration` and `TestEndpointProvider_ReturnsSameSequencerOnInactiveWithLongCheckDuration` cover this case, no? Or do I misunderstand?

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -13,6 +13,10 @@ const DefaultActiveSequencerFollowerCheckDuration = 2 * DefaultDialTimeout
+// ActiveL2EndpointProvider is an interface for providing a RollupClient and l2 eth client
+// It manages the lifecycle of the RollupClient and eth client for callers
+// It does this by failing over down the list of rollupUrls if the current one is inactive or broken
+
```
```suggestion
```

Commit:
op-service: obey linter in rabbit-provided error message change.

EvanJRichard commented:
Sidenote: flaky test has occurred in `TestSend/dual_threaded_blocking`. Will spend time today investigating flaky test, and log results in issue tracker.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -61,26 +67,22 @@ func newActiveL2RollupProvider(
+	if err != nil {
+		return nil, fmt.Errorf("setting provider rollup client: %w", err)
+	}
+	return p, nil
```
If the sequencer with index 0 is offline at startup, the dialing will error and the ctor return that error. We should add a test for this case, seems we only have a test right now that an _inactive_ seq for index 0 still works, but not an _offline_ one. This was also the case when the nil-case dialing was still in `RollupClient()`.

There should be a way to already run the activity check loop at creation without duplicating the entire logic here. Maybe one way is to simply let `checkCurrentSequencer` return `false` if the current one is nil, and setting the field `currentRollupClient` to `-1` further up in the ctor when creating the `ActiveL2RollupProvider` so that the first index that is dialed is then 0.

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -79,8 +93,8 @@ func (p *ActiveL2EndpointProvider) EthClient(ctx context.Context) (EthClientInte
 	if err != nil {
 		return nil, err
 	}
-	if p.ethClientIndex != p.rollupIndex || p.currentEthClient == nil {
```
I see that the logic now got mostly duplicated into the ctor, but couldn't be removed here. As it was before was then probably better after all. Sorry didn't realize it would turn out like this.

EvanJRichard replied:
No worries, I like it either way. I'll go back to the old way.

sebastianst reviewed (changes requested):
I realize now, don't we need to call `AssertExpectations` at the end of each test on the mock(s)?

EvanJRichard committed:
Update op-service/dial/active_l2_provider.go

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

Commit:
op-service active L2 provider tests: assertAllExpectations after most tests.

Commit:
op-service: more elegantly handle startup in active l2 providers, and improve testing.

Commit:
Merge branch 'evan/batcher-active-seq' of github.com:ethereum-optimism/optimism into evan/batcher-active-seq

Commit:
Change remaining longDurationTests to be able to use ept.assertAllExpectations.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -105,10 +100,23 @@ func (p *ActiveL2RollupProvider) shouldCheck() bool {
 		active, err := p.checkCurrentSequencer(ctx)
-		ep := p.rollupUrls[p.rollupIndex]
+		ep := p.currentEndpoint()
 		if err != nil {
```
I see what happens here. We only need the endpoint string in logs. What we could do instead, is have `checkCurrentSequencer()` return `false, errSeqUnset` if the current one is `nil`, where `var errSeqUnset = errors.New("sequencer unset")`, and then do here

```
if errors.Is(err, errSeqUnset) {
    log.Debug("Current sequencer unset.")
} else if ep := p.rollupUrls[p.rollupIndex]; err != nil {
// continue remaining as it is
```

This has the advantage that
- we can remove the somewhat odd `currentEndpoint` function,
- the logs are even clearer,
- we can remove the duplicate dial above the loop

EvanJRichard replied:
The way this is currently structured, we still need the duplicate dial, or else tests like `TestEndpointProvider_SelectSecondSequencerIfFirstInactiveAtCreation` will fail. The provider successfully dials the second sequencer, but then the loop is exhausted, so it doesn't `checkCurrentSequencer`. Looking into how to restructure the loop to handle that case.

EvanJRichard replied:
Made a readability change but the situation stands: with the new design, where `p.currentRollupClient` is initially `nil` and then initialized through a call to `RollupClient()`, we need to call `dialNextSequencer` one time more than the loop dials. But, this extra dial only needs to occur in the case where `p.currentRollupClient` is `nil`. Maybe we move it to the constructor? Hmm.

Added a many-sequencers, only-last-is-active pair of tests, too. They pass with the duplicate `dial`, and fail without it.

EvanJRichard replied:
Moved the one-time-extra-dial to the constructor, it's a bit cleaner there, I think.

Commit:
use new error errSeqUnset.

Commit:
Add test for scenario where many sequencers are inactive, and only the last is active.

Commit:
Readability change: move the on-creation initialization to its own function.

Commit:
Move extra one-time dial to constructor.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,169 @@
+	}
+	cctx, cancel := context.WithTimeout(ctx, networkTimeout)
+	defer cancel()
+	err := p.ensureClientInitialized(cctx)
+	if err != nil {
+		return nil, fmt.Errorf("dialing initial rollup client: %w", err)
+	}
```
Doesn't this still have the problem that if the first sequencer is not only inactive, but dead (= dialing fails), then the ctor errors? We want the ctor to survive offline sequencers and find the one active online sequencer.

EvanJRichard commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,169 @@
+
+func (p *ActiveL2RollupProvider) Close() {
+	p.currentRollupClient.Close()
+}
```
Better safe than sorry, sure.

EvanJRichard committed:
Update op-service/dial/active_rollup_provider.go

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,169 @@
+		} else {
+			p.log.Info("Current sequencer inactive, closing connection and trying next.", "index", p.rollupIndex, "url", ep)
+		}
+		if err := p.dialNextSequencer(ctx); err != nil {
+			return fmt.Errorf("dialing next sequencer: %w", err)
+		}
```
I think we want to do
```suggestion
		if err := p.dialNextSequencer(ctx); err != nil {
			p.log.Warn("Error dialing next sequencer.", "err", err, "index", p.rollupIndex)
		}
```
here and exhaustively iterating over the whole loop. If dialing fails, we also want to try the next sequencer.

I think if we do that, we might be able to remove `ensureClientInitialized` and also not dial from the ctor.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,683 @@
+
+// TestRollupProvider_SelectLastSequencerIManyInactiveAtCreation verifies that if all but the last sequencer
+// are inactive at the time of ActiveL2RollupProvider creation, the last active sequencer is chosen.
+func TestRollupProvider_SelectLastSequencerIfManyInactiveAtCreation(t *testing.T) {
```
We also want a test where the first 4 sequencers are offline, i.e., not dialable. We might be able to just adjust the test for 2. No need to duplicate test for 2 and 5 seqs.

EvanJRichard replied:
changed `ManyInactive` to `ManyOffline` and changed test flow to make them not dialable rather than errored.

Commit:
Add nil check to active l2 provider.

sebastianst reviewed (commented):
I may be confused, don't we want to retry a new sequencer if dialing one fails?

We might need to add functionality to the test setup to let the dialers error on dial.

EvanJRichard committed:
Update op-service/dial/active_rollup_provider.go

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

EvanJRichard commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,112 @@
+
+const DefaultActiveSequencerFollowerCheckDuration = 2 * DefaultDialTimeout
+
+type ethDialer func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (EthClientInterface, error)
```
What do you mean, doesn't `NewActiveL2EndpointProvider` use a `Context`, two `[]string`, two `Duration`, and a `log.Logger`? `newActiveL2EndpointProvider` uses an `ethDialer` parameter, but it's non-exported.

EvanJRichard commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,112 @@
+	ethUrls          []string
+}
+
+// NewActiveL2EndpointProvider creates a new ActiveL2EndpointProvider
+// the checkDuration is the duration between checks to see if the current rollup client is active
+// provide a checkDuration of 0 to check every time
+func NewActiveL2EndpointProvider(ctx context.Context,
+	ethUrls, rollupUrls []string,
+	checkDuration time.Duration,
+	networkTimeout time.Duration,
+	logger log.Logger,
+) (*ActiveL2EndpointProvider, error) {
+	ethDialer := func(ctx context.Context, timeout time.Duration,
+		log log.Logger, url string) (EthClientInterface, error) {
+		return DialEthClientWithTimeout(ctx, timeout, log, url)
+	}
+	rollupDialer := func(ctx context.Context, timeout time.Duration,
+		log log.Logger, url string) (RollupClientInterface, error) {
+		return DialRollupClientWithTimeout(ctx, timeout, log, url)
+	}
+	return newActiveL2EndpointProvider(ctx, ethUrls, rollupUrls, checkDuration, networkTimeout, logger, ethDialer, rollupDialer)
```
Show me how to inline them as suggested, please, I must misunderstand.

EvanJRichard replied:
There's a misunderstanding, the functions are already inlined as you described.

Commit:
Address review comment: change many-inactive tests to many-undialable tests.

sebastianst committed:
op-service: Improve active seq provider

- Preserve the invariant that the index and current rollup/eth
  client match.
- Dial at the start of the loop instead of at the end.

Commit:
Add test that reproduces internal state corruption.

Commit:
Fix some tests.

Commit:
Move usage of ExpectClose to MaybeClose, we don't want to enforce a particular close behavior in these tests.

Commit:
add a missing call to assertAllExpectations.

EvanJRichard committed:
Merge pull request #8673 from ethereum-optimism/seb/batcher-active-seq-1

op-service: Improve active seq provider

EvanJRichard commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,114 @@
+	return p.currentEthClient, nil
+}
+
+func (p *ActiveL2EndpointProvider) Close() {
+	if p.currentEthClient != nil {
+		p.currentEthClient.Close()
+	}
+	p.ActiveL2RollupProvider.Close()
+}
```
Good thought, but `Close()` doesn't return an `err`.

EvanJRichard commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,114 @@
+	return p, nil
+}
+
+func (p *ActiveL2EndpointProvider) EthClient(ctx context.Context) (EthClientInterface, error) {
+	p.clientLock.Lock()
+	defer p.clientLock.Unlock()
+	err := p.ensureActiveEndpoint(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if p.ethClientIndex != p.rollupIndex || p.currentEthClient == nil {
+		// we changed sequencers, dial a new EthClient
+		cctx, cancel := context.WithTimeout(ctx, p.networkTimeout)
+		defer cancel()
+		idx := p.rollupIndex
+		ep := p.ethUrls[idx]
+		log.Info("sequencer changed (or ethClient was nil due to startup), dialing new eth client", "new_index", idx, "new_url", ep)
+		ethClient, err := p.ethDialer(cctx, p.networkTimeout, p.log, ep)
+		if err != nil {
+			return nil, fmt.Errorf("dialing eth client: %w", err)
+		}
+		if p.currentEthClient != nil {
+			p.currentEthClient.Close()
+		}
+		p.ethClientIndex = idx
+		p.currentEthClient = ethClient
+	}
+	return p.currentEthClient, nil
```
Good proposal, but I think it's worth keeping the easier-to-reason-about lock behavior in exchange for a potentially long lock hold time.

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,744 @@
+	seq1.MaybeClose()
+	// tertiarySequencer can't even be dialed
+	ept.setRollupDialOutcome(2, false)
+	// after calling RollupClient, index will be set to 0, but currentRollupClient is secondarySequencer
```
this doesn't apply any more, was only true during debugging of original buggy impl. 
we could instead briefly document the old bug here, like "In a prior buggy implementation, this scenario lead to an internal inconsistent state where the current client didn't match the index. On a subsequent try, this lead to the active sequencer at 0 to be skipped entirely, while the sequencer at index 1 was checked twice."

EvanJRichard replied:
Will use exactly this phrasing :+1:

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,744 @@
+	rollupClient, err := rollupProvider.RollupClient(context.Background())
+	require.Error(t, err)
+	require.Nil(t, rollupClient)
+	// internal state would now be inconsistent in the buggy impl.
```
```suggestion
	// internal state would now be inconsistent in a buggy impl.
```
since it's now more general about any would-be buggy impl

sebastianst commented on op-service/dial/active_l2_provider_test.go:
```diff
@@ -0,0 +1,744 @@
+	ept.setRollupDialOutcome(2, true)
+	seq2.ExpectSequencerActive(false, nil)
+	seq2.MaybeClose()
+	// now trigger the bug by calling for the client
```
there isn't a bug any more.

sebastianst commented on op-service/dial/active_rollup_provider.go:
```diff
@@ -0,0 +1,169 @@
+	logger log.Logger,
+	dialer rollupDialer,
+) (*ActiveL2RollupProvider, error) {
+	if len(rollupUrls) == 0 {
```
I wondering, do we have any test that asserts (special) behavior in the edge case where we have only a single url?

EvanJRichard replied:
If we have only a single url, we should be using the `StaticL2RollupProvider`/`StaticL2EndpointProvider` - nonetheless, it'd be good for the `Active` one to handle this scenario if it were somehow to occur. Adding such a test.

EvanJRichard replied:
Added `TestRollupProvider_HandlesSingleSequencer` and `TestEndpointProvider_HandlesSingleSequencer`

sebastianst commented on op-service/dial/active_l2_provider.go:
```diff
@@ -0,0 +1,114 @@
+	if err != nil {
+		return nil, err
+	}
+	if p.ethClientIndex != p.rollupIndex || p.currentEthClient == nil {
```
do we have a simpe test that asserts that if we call `EthClient` twice for the same active sequencer, it then just returns the current one and doesn't enter this function to re-dial?

EvanJRichard replied:
Yes! `TestEndpointProvider_LongCheckDuration` around line 584 in the test file

sebastianst reviewed (commented):
Should be good to go! There are some comments to fix in the regression test. Also proposed another 1-2 tests that we might have missed.

Commit:
Test even the case where the active providers are managing a list of 1 element.

Commit:
Revert experimental hunk in active_l2_provider.

sebastianst reviewed (approved):
LGTM ✨
